barman/recovery_executor.py

```diff
@@ -1447,13 +1462,293 @@ def _backup_copy(
+                mount_options_errors.append(
+                    "Device %s cloned from snapshot %s is mounted with %s but %s was "
+                    "expected."
+                    % (device, snapshot, mount_options, expected_mount_options)
+                )
+        if len(mount_point_errors) > 0:
+            raise RecoveryPreconditionException(
+                "Error checking mount points: %s" % ", ".join(mount_point_errors)
+            )
+        if len(mount_options_errors) > 0:
+            raise RecoveryPreconditionException(
+                "Error checking mount options: %s" % ", ".join(mount_options_errors)
+            )
+
+    def recover(
+        self,
+        backup_info,
+        dest,
+        tablespaces=None,
+        remote_command=None,
+        target_tli=None,
+        target_time=None,
+        target_xid=None,
+        target_lsn=None,
+        target_name=None,
+        target_immediate=False,
+        exclusive=False,
+        target_action=None,
+        standby_mode=None,
+        recovery_instance=None,
+        recovery_zone=None,
+    ):
+        """
+        Performs a recovery of a snapshot backup.
+
+        This method should be called in a closing context.
+
+        :param barman.infofile.BackupInfo backup_info: the backup to recover
+        :param str dest: the destination directory
+        :param dict[str,str]|None tablespaces: a tablespace
+            name -> location map (for relocation)
+        :param str|None remote_command: The remote command to recover
+                               the base backup, in case of remote backup.
+        :param str|None target_tli: the target timeline
+        :param str|None target_time: the target time
+        :param str|None target_xid: the target xid
+        :param str|None target_lsn: the target LSN
+        :param str|None target_name: the target name created previously with
+                            pg_create_restore_point() function call
+        :param str|None target_immediate: end recovery as soon as consistency
+            is reached
+        :param bool exclusive: whether the recovery is exclusive or not
+        :param str|None target_action: The recovery target action
+        :param bool|None standby_mode: standby mode
+        :param str|None recovery_instance: The name of the recovery node as it is
+            known by the cloud provider
+        :param str|None recovery_zone: The zone in which the recovery node is located
+        """
+        snapshot_interface = get_snapshot_interface_from_backup_info(backup_info)
+        attached_snapshots = self.get_attached_snapshots_for_backup(
+            snapshot_interface, backup_info, recovery_instance, recovery_zone
+        )
+        cmd = fs.unix_command_factory(remote_command, self.server.path)
+        SnapshotRecoveryExecutor.check_mount_points(
+            backup_info, attached_snapshots, cmd
+        )
+        self.check_recovery_dir_exists(dest, cmd)
```

Comment:
I wonder if we should also be checking ownership and permissions of the recovery directory (and any tablespace directories). The reason being that, unless the recovery instance is provisioned exactly the same way as the original backup instance, the postgres user and group may have different uid and gids on the recovery instance. If that happens then the files on the disks cloned from the snapshots will not belong to the postgres user on the recovery instance.

In such cases we won't notice until we try and interact with the config files during Barman's part of the recovery process. At that point we will fail noisily and it *should* be clear what the problem is, but it would be nicer if Barman detected it here and failed with a nice human-readable warning.

Reply:
This would be a good future enhancement but I don't think it is necessary for the first iteration of snapshot backups since we already fail noisily in this scenario.